src/FileInput/FileInput.jsx

```diff
@@ -0,0 +1,136 @@
+    </g>
+  </svg>);
+}
+
+function UploadingIcon(props) {
+  return (<svg className="FileInput--Icon" width="42px" height="24px" viewBox="0 0 42 24">
+    <linearGradient id="bg" x1="0.5" y1="1" x2="0.5" y2="0">
+      <stop offset="0%" stopOpacity="1" stopColor="#4274f6" />
+      <stop offset={`${props.percent}%`} stopOpacity="1" stopColor="#4274f6" />
+      <stop offset={`${props.percent}%`} stopOpacity="1" stopColor="#E3E6EB" />
+      <stop offset="100%" stopOpacity="1" stopColor="#E3E6EB" />
+    </linearGradient>
+    <g>
+      <g transform="translate(-242.000000, -13.000000)" fill="url(#bg)">
+        <path d="M248.461538,23.1333333 C243.615385,23.1333333 242,26.8666667 242,29.6715179 C242,33.2356824 244.153846,36.4622091 248.461538,37 C255,37 273,37 277.269231,37 C280.689148,37 284,33.0511418 284,28.5959361 C284,24.1407304 280.230769,19.9912817 275.922774,20.7203129 C275.884815,15.8553426 272.28333,13 267.846154,13 C263,13 260.307692,18.3333333 260.307692,18.3333333 C260.307692,18.3333333 258.692308,16.7333333 255.461538,16.7333333 C249,16.7333333 248.461538,23.1333333 248.461538,23.1333333 Z"></path>
+      </g>
+    </g>
+  </svg>);
+}
+
+UploadingIcon.propTypes = {
+  percent: React.PropTypes.number.isRequired,
+};
+
+export class FileInput extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      filename: null,
+      error: null,
+      progress: null,
+      success: null,
+    };
+    this.onDropRejected = this.onDropRejected.bind(this);
+    this.onDropAccepted = this.onDropAccepted.bind(this);
+  }
+
+  onDropAccepted(files) {
+    const fileKey = Math.random();
+    this.setState({fileKey});
+    this.setState({filename: files[0].name, error: null, progress: 0, success: null});
+    this.props.store(files[0], {
+      success: () => {
+        if (this.state.fileKey !== fileKey) {
+          return; // If the user drops multiple files we want to only treat the last one as "success"
+        }
+        this.setState({success: true});
+      },
+      error: () => {
+        this.setState({error: "Unknown error"});
+      },
+      progress: (progress) => {
+        if (this.state.fileKey !== fileKey) {
+          return; // If the user drops multiple files we want to only take progress from the active upload
+        }
+        this.setState({progress});
+      },
+    });
+  }
+
+  onDropRejected() {
+    this.setState({filename: null, error: `Invalid ${this.props.label}`, success: null});
+  }
+
+  render() {
+    return (<Dropzone
+      accept={this.props.accept}
+      style={{}}
+      multiple={false}
+      onDropAccepted={this.onDropAccepted}
+      onDropRejected={this.onDropRejected}
+    >{({isDragActive, isDragReject}) => {
```

Comment:
It's basically making a component inline is how I understood it, but yea, interesting.